htlcswitch/quiescer.go

```diff
@@ -0,0 +1,233 @@
+//nolint:unused
+package htlcswitch
+
+import (
+	"fmt"
+
+	"github.com/lightningnetwork/lnd/fn"
+	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+var (
+	// ErrStfuAlreadySent indicates that this channel has already sent an
+	// Stfu message for this negotiation.
+	ErrStfuAlreadySent = fmt.Errorf("stfu already sent")
+
+	// ErrStfuAlreadyRcvd indicates that this channel has already received
+	// an Stfu message for this negotiation.
+	ErrStfuAlreadyRcvd = fmt.Errorf("stfu already received")
+
+	// ErrNoQuiescenceInitiator indicates that the caller has requested the
+	// quiescence initiator for a channel that is not yet quiescent.
+	ErrNoQuiescenceInitiator = fmt.Errorf(
+		"indeterminate quiescence initiator: channel is not quiescent",
+	)
+
+	// ErrPendingRemoteUpdates indicates that we have received an Stfu while
+	// the remote party has issued updates that are not yet bilaterally
+	// committed.
+	ErrPendingRemoteUpdates = fmt.Errorf(
+		"stfu received with pending remote updates",
+	)
+
+	// ErrPendingLocalUpdates indicates that we are attempting to send an
+	// Stfu while we have issued updates that are not yet bilaterally
+	// committed.
+	ErrPendingLocalUpdates = fmt.Errorf(
+		"stfu send attempted with pending local updates",
+	)
+)
+
+type quiescerCfg struct {
+	// chanID marks what channel we are managing the state machine for. This
+	// is important because the quiescer is responsible for constructing the
+	// messages we send out and the ChannelID is a key field in that
+	// message.
+	chanID lnwire.ChannelID
+
+	// channelInitiator indicates which ChannelParty originally opened the
+	// channel. This is used to break ties when both sides of the channel
+	// send Stfu claiming to be the initiator.
+	channelInitiator lntypes.ChannelParty
+
+	// numPendingUpdates is a function that returns the number of pending
+	// originated by the party in the first argument that have yet to be
+	// committed to the commitment transaction held by the party in the
+	// second argument.
+	numPendingUpdates func(lntypes.ChannelParty,
+		lntypes.ChannelParty) uint64
+
+	// sendMsg is a function that can be used to send an Stfu message over
+	// the wire.
+	sendMsg func(lnwire.Stfu) error
+}
+
+// quiescer is a state machine that tracks progression through the quiescence
+// protocol.
+type quiescer struct {
+	cfg quiescerCfg
+
+	// localInit indicates whether our path through this state machine was
+	// initiated by our node. This can be true or false independently of
+	// remoteInit.
+	localInit bool
+
+	// remoteInit indicates whether we received Stfu from our peer where the
+	// message indicated that the remote node believes it was the initiator.
+	// This can be true or false independently of localInit.
+	remoteInit bool
```

Comment:
Drive-by commentary (non-blocking): another way to structure this would have been to use distinct structs/states for each of the logical packages of these attributes. So `WaitingForStfu`, `StfuSent`, `SftuActive`, etc. That way a reader doesn't need to keep track of which set of attributes maps to a given state, they instead just see the state with the set of relevant attributes. 

There's also be no mutation in this case, just one state accepting w/e event and creating another state.

Reply:
Yeah there are a number of ways that this state machine could be reimagined. That said, since either side can initiate the process, the whole state machine needs to accommodate a mirrored version. Should be straightforward to use ChannelParty for this. I'll probably go this route for the dyncomm state machine.